core/sugartypes.ml

```diff
@@ -73,36 +76,140 @@ let default_effect_subkind : Subkind.t = (lin_unl, res_any)
     [@@deriving show]
 
 type quantifier = type_variable
   [@@deriving show]
 
+
+module SugarTypeVar =
+struct
+
+
+(* Note that an unresolved type variable does not contain information
+   about its primary kind. This is filled in when resolving the variable *)
+type t =
+  | TUnresolved       of Name.t * Subkind.t option * Freedom.t
+  (* This is why we can't have nice things ... *)
+  | TResolvedType     of Types.meta_type_var
+  | TResolvedRow      of Types.meta_row_var
+  | TResolvedPresence of Types.meta_presence_var
+     [@@deriving show]
+
+
+let is_resolved = function
+  | TUnresolved _ -> false
+  | _ -> true
+
+let mk_unresolved name subkind_opt freedom_opt =
+  TUnresolved (name, subkind_opt, freedom_opt)
+
+let mk_resolved_tye point : t =
+  TResolvedType point
+
+let mk_resolved_row point : t =
+  TResolvedRow point
+
+let mk_resolved_presence point : t =
+  TResolvedPresence point
+
+
+let get_unresolved_exn = function
+  | TUnresolved (name, subkind, freedom) ->
+     name, subkind, freedom
+  | _ ->
+     raise
+       (internal_error
+          "Requesting unresolved type var when
+           it has already been resolved")
+
+let get_unresolved_name_exn =
+  get_unresolved_exn ->- fst3
+
+ let get_resolved_type_exn =
+   function
+   | TResolvedType point -> point
+   | _ -> raise (internal_error "requested kind does not match existing kind info")
+
+let get_resolved_row_exn =
+  function
+  | TResolvedRow point -> point
+  | _ -> raise (internal_error "requested kind does not match existing kind info")
+
+let get_resolved_presence_exn =
+  function
+  | TResolvedPresence point -> point
+  | _ -> raise (internal_error "requested kind does not match existing kind info")
+
+
+end
+
+
+module SugarQuantifier =
+struct
+
+  type t =
+    | QUnresolved of Name.t * kind * Freedom.t
+    | QResolved of Quantifier.t
+      [@@deriving show]
+
+
+  let mk_unresolved name kind freedom =
+    QUnresolved (name, kind, freedom)
+
+  let mk_resolved quantifier =
+    QResolved quantifier
+
+
+  let get_unresolved_exn = function
+    | QUnresolved (name, kind, freedom) -> name, kind, freedom
+    | QResolved _q ->
+       raise
+         (internal_error
+            "Requesting unresolved quantifier when
+             it has already been resolved")
+
+ let get_unresolved_name_exn =
+   get_unresolved_exn ->- fst3
+
+
+  let get_resolved_exn = function
+    | QResolved q -> q
+    | QUnresolved _ ->
+       raise
+         (internal_error
+            "Requesting resolved type var before
+             it has been resolved")
+
+end
+
+
+
 let rigidify (name, kind, _) = (name, kind, `Rigid)
 
-let string_of_type_variable ((var, (kind, subkind), _) : type_variable) =
-  match kind with
-  | None -> var
-  | Some kind ->
-     let subkind = OptionUtils.opt_app Subkind.to_string "" subkind in
-     var ^ "::" ^ PrimaryKind.to_string kind ^ subkind
+let string_of_type_variable (_v : SugarTypeVar.t) =
+  failwith "fixme"
```

Comment:
What would it mean to fix this?  Is this function still used anywhere?  It doesn't appear in the diff.

Reply:
Ah, good point. I disabled this function and used an auto-generated printer in the one place where this function was originally used. I've now restored this (nicer) version and given it a proper implementation.